### PR TITLE
successful merge banner

### LINKED ISSUE #5851: Lack of feedback after merge is completed
This came out of research in #5400, and hopefully should be released along with it. 

We hope this can act as a proof of concept for feedback from other interactions in Desktop, where if this is well received and helpful for people, we'll be able to utilize it in other places in the app.

**Please describe the problem you think should be solved**

After I complete a merge, it's difficult to tell what happened or whether something happened successfully. It'd be helpful to have some sort of visual feedback to let me know the merge completed successfully. This is relevant both in a merge without conflicts and with a merge that is completed after resolving conflicts in the new flow from #5400. 

**[Optional] Do you have any potential solutions in mind?**

A couple considerations:

1. We already have the concept of a small banner under the Current Repo, Current Branch, and Fetch/Pull/Push buttons for when updates are released.

2. We've heard on multiple occasions that the post-commit and post-merge flow is unhelpful because it's hard to remember that you likely want to push immediately afterward. Some kind of visual representation on the "Push" button might be helpful to nudge people toward that as their next action.

cc: @desktop/design @outofambit

### PR DESCRIPTION
**Pull request by outofambit**
## Overview

**Closes #5851**

Adds a banner that appears when the user completes a merge.

screenshot:

success banner

in action:

great success

## Description

- This PR will need to be updated once #5809 is merged to make sure we cover all ways of merging
- This required more reworking of the git-related parts of the codebase than i would have liked, so if there alternate approaches, I'm all ears
- This PR _doesn't_ attempt to make a banner management system as we only have two banners currently. (As we add more we'll have to cross that bridge.)

## Release notes

Notes: Displays merge success banner

### CODE REVIEW

**Comment by agisilaos** [comment]
@outofambit Since this is a UI change can you provide a screenshot in the description about how the banner is gonna look like? It would be easier to see the final result without going back and forth in the issue or run the app locally.

**Comment by outofambit** [comment]
@agisilaos good point! added now :)

File: app/src/lib/app-state.ts [hunk 1]
```diff
@@ -526,3 +529,10 @@ export interface ICompareToBranch {
  * An action to send to the application store to update the compare state
  */
 export type CompareAction = IViewHistory | ICompareToBranch
+
+export type SuccessfulMergeBannerState =
+  | {
+      currentBranch: string
+      theirBranch: string
+    }
+  | false
```
Comment (shiftkey) [inline app/src/lib/app-state.ts]: :art: could we use `null` here to indicate the lack of successful merge banner state? It feels weird here to be mashing together an object with a flag.

File: app/src/lib/stores/git-store.ts [hunk 2]
```diff
@@ -1114,7 +1114,7 @@ export class GitStore extends BaseStore {
   }
 
   /** Merge the named branch into the current branch. */
-  public merge(branch: string): Promise<void> {
+  public merge(branch: string): Promise<void | true> {
```
Comment (shiftkey) [inline app/src/lib/stores/git-store.ts]: I'm wrapping my head around these internals changed but I have some suggestions for this code path:

 - omit `true` from the inner `Promise<T>` return type on the Git operation (if the action fails, the promise is rejected and that needs to be handled):

```diff
diff --git a/app/src/lib/git/merge.ts b/app/src/lib/git/merge.ts
index bccd9ee45..2e78d44b4 100644
--- a/app/src/lib/git/merge.ts
+++ b/app/src/lib/git/merge.ts
@@ -9,9 +9,8 @@ import { spawnAndComplete } from './spawn'
 export async function merge(
   repository: Repository,
   branch: string
-): Promise<void | true> {
+): Promise<void> {
   await git(['merge', branch], repository.path, 'merge')
-  return true
 }

 /**
```

 - change `GitStore` return a `boolean` to indicate whether the operation succeeded or not (internal error handling takes care of the rest):

```diff
diff --git a/app/src/lib/stores/git-store.ts b/app/src/lib/stores/git-store.ts
index 598416e30..cec479f93 100644
--- a/app/src/lib/stores/git-store.ts
+++ b/app/src/lib/stores/git-store.ts
@@ -1114,10 +1114,15 @@ export class GitStore extends BaseStore {
   }
 
   /** Merge the named branch into the current branch. */
-  public merge(branch: string): Promise<void | true> {
-    return this.performFailableOperation(() => merge(this.repository, branch), {
-      command: 'merge',
-    })
+  public async merge(branch: string): Promise<boolean> {
+    const result = await this.performFailableOperation(
+      () => merge(this.repository, branch),
+      {
+        command: 'merge',
+      }
+    )
+
+    return result !== undefined
   }
 
   /** Changes the URL for the remote that matches the given name  */
```
 
 - tidy up the `AppStore` checks (only show the dialog when you have a valid tip):

```diff
diff --git a/app/src/lib/stores/app-store.ts b/app/src/lib/stores/app-store.ts
index ffd92c175..b39108954 100644
--- a/app/src/lib/stores/app-store.ts
+++ b/app/src/lib/stores/app-store.ts
@@ -3079,13 +3079,12 @@ export class AppStore extends TypedBaseStore<IAppState> {
       }
     }
 
-    if ((await gitStore.merge(branch)) === true) {
-      const currentBranchName =
-        gitStore.tip.kind === TipState.Valid
-          ? gitStore.tip.branch.name
-          : 'unknown branch'
+    const mergeSuccessful = await gitStore.merge(branch)
+    const { tip } = gitStore
+
+    if (mergeSuccessful && tip.kind === TipState.Valid) {
       this._setSuccessfulMergeBannerState({
-        currentBranch: currentBranchName,
+        currentBranch: tip.branch.name,
         theirBranch: branch,
       })
     }
```

Comment (outofambit) [reply app/src/lib/stores/git-store.ts]:
```diff
+  public async merge(branch: string): Promise<boolean> {
+    const result = await this.performFailableOperation(
+      () => merge(this.repository, branch),
+      {
+        command: 'merge',
+      }
+    )
+
+    return result !== undefined
```

@shiftkey so i tried something like this first, but the part i got stuck on was differentiating a `void` return and an `undefined` return. i think the same issue is here in this implementation?

> tidy up the AppStore checks (only show the dialog when you have a valid tip):

👍 will make this change

Comment (shiftkey) [reply app/src/lib/stores/git-store.ts]: Possibly? I'll think about this some more and see if I can spot a non-ugly way out of this...

Comment (outofambit) [reply app/src/lib/stores/git-store.ts]: @shiftkey yep that sounds about right

**Comment by outofambit** [comment]
thank you @shiftkey!

**Comment by tierninho** [comment]
Tested and LGTM 👍

**Comment by shiftkey** [comment]
Ignore that pending merge status, it totally went green :shipit:

**Comment by tierninho** [comment]
Spotted this on 1.5.0-beta2 after selecting Branch > Update from <branch_name>

screen shot 2018-10-30 at 1 25 56 pm

Truncation needed 😄

**Comment by billygriffin** [comment]
It doesn't seem like this shows the success banner after you've completed a merge successfully after resolving conflicts. @donokuda and I both tested it on beta3, and it showed up after a normal merge, but not after a conflicted merge was completed successfully.

**Pull request merged**